File: java/src/main/java/com/shipzy/sdk/EpodClient.java
package com.shipzy.sdk;

import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.Duration;
import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;
import com.google.gson.Gson;
import com.google.gson.annotations.SerializedName;
import java.util.List;
import java.util.Map;

public class EcmrClient {
    private final HttpClient httpClient;
    private final EpodClient.ShipzyConfig config;
    private static final Gson gson = new Gson();

    public EcmrClient(EpodClient.ShipzyConfig c) {
        this.config = c;
        this.httpClient = HttpClient.newBuilder().connectTimeout(Duration.ofSeconds(c.getTimeoutSeconds())).build();
    }
    public void setToken(String t) { config.setToken(t); }

    private <T> T request(String path, String method, Class<T> t, Object body) throws Exception {
        var builder = HttpRequest.newBuilder()
            .uri(URI.create(config.getBaseUrl().replaceAll("/$", "") + path))
            .header("Authorization", new EpodClient(config).getAuthHeader())
            .header("Content-Type", "application/json");
        if ("POST".equals(method) || "PUT".equals(method))
            builder.method(method, HttpRequest.BodyPublishers.ofString(body != null ? gson.toJson(body) : "{}"));
        else builder.GET();
        HttpResponse<String> resp = httpClient.send(builder.build(), HttpResponse.BodyHandlers.ofString());
        if (resp.statusCode() == 401) throw new EpodClient.ShipzyAuthException("Unauthorized");
        if (resp.statusCode() >= 400) throw new EpodClient.ShipzyException("HTTP " + resp.statusCode(), resp.statusCode());
        return gson.fromJson(resp.body(), t);
    }

    public Object list(int page, int pageSize) throws Exception {
        return request("/api/v1/shipment/ecmr/list?page=" + page + "&page_size=" + pageSize, "GET", Object.class, null);
    }
    public Object get(String id) throws Exception { return request("/api/v1/shipment/ecmr/" + id, "GET", Object.class, null); }
    public Object create(Object data) throws Exception { return request("/api/v1/shipment/ecmr/create", "POST", Object.class, data); }
    public Object generateFromOrder(String orderId) throws Exception { return request("/api/v1/shipment/ecmr/generate-from-order", "POST", Object.class, Map.of("order_id", orderId)); }
    public Object sign(String id) throws Exception { return request("/api/v1/shipment/ecmr/" + id + "/sign", "POST", Object.class, null); }
    public Object pdf(String id) throws Exception { return request("/api/v1/shipment/ecmr/" + id + "/pdf", "POST", Object.class, null); }
}

class ProductClient {
    private final HttpClient httpClient;
    private final EpodClient.ShipzyConfig config;
    private static final Gson gson = new Gson();

    public ProductClient(EpodClient.ShipzyConfig c) {
        this.config = c;
        this.httpClient = HttpClient.newBuilder().connectTimeout(Duration.ofSeconds(c.getTimeoutSeconds())).build();
    }
    public void setToken(String t) { config.setToken(t); }

    private <T> T request(String path, String method, Class<T> t, Object body) throws Exception {
        var builder = HttpRequest.newBuilder()
            .uri(URI.create(config.getBaseUrl().replaceAll("/$", "") + path))
            .header("Authorization", new EpodClient(config).getAuthHeader())
            .header("Content-Type", "application/json");
        if ("POST".equals(method) || "PUT".equals(method))
            builder.method(method, HttpRequest.BodyPublishers.ofString(body != null ? gson.toJson(body) : "{}"));
        else builder.GET();
        HttpResponse<String> resp = httpClient.send(builder.build(), HttpResponse.BodyHandlers.ofString());
        if (resp.statusCode() == 401) throw new EpodClient.ShipzyAuthException("Unauthorized");
        if (resp.statusCode() >= 400) throw new EpodClient.ShipzyException("HTTP " + resp.statusCode(), resp.statusCode());
        return gson.fromJson(resp.body(), t);
    }

    public Object list(String status, String category, String search, Boolean activeOnly) throws Exception {
        StringBuilder q = new StringBuilder("?active_only=true");
        if (status != null) q.append("&status=").append(URLEncoder.encode(status, StandardCharsets.UTF_8));
        if (category != null) q.append("&category=").append(URLEncoder.encode(category, StandardCharsets.UTF_8));
        if (search != null) q.append("&search=").append(URLEncoder.encode(search, StandardCharsets.UTF_8));
        if (activeOnly != null) q.append("&active_only=").append(activeOnly);
        return request("/api/v1/products" + q, "GET", Object.class, null);
    }
    public Object get(String id) throws Exception { return request("/api/v1/products/" + id, "GET", Object.class, null); }
    public Object create(Object data) throws Exception { return request("/api/v1/products", "POST", Object.class, data); }
    public Object update(String id, Object data) throws Exception { return request("/api/v1/products/" + id, "PUT", Object.class, data); }
    public Object retire(String id) throws Exception { return request("/api/v1/products/" + id + "/retire", "POST", Object.class, null); }
}

class ActivationClient {
    private final HttpClient httpClient;
    private final EpodClient.ShipzyConfig config;
    private static final Gson gson = new Gson();

    public ActivationClient(EpodClient.ShipzyConfig c) {
        this.config = c;
        this.httpClient = HttpClient.newBuilder().connectTimeout(Duration.ofSeconds(c.getTimeoutSeconds())).build();
    }
    public void setToken(String t) { config.setToken(t); }

    private <T> T request(String path, String method, Class<T> t, Object body) throws Exception {
        var builder = HttpRequest.newBuilder()
            .uri(URI.create(config.getBaseUrl().replaceAll("/$", "") + path))
            .header("Authorization", new EpodClient(config).getAuthHeader())
            .header("Content-Type", "application/json");
        if ("POST".equals(method) || "PUT".equals(method))
            builder.method(method, HttpRequest.BodyPublishers.ofString(body != null ? gson.toJson(body) : "{}"));
        else builder.GET();
        HttpResponse<String> resp = httpClient.send(builder.build(), HttpResponse.BodyHandlers.ofString());
        if (resp.statusCode() == 401) throw new EpodClient.ShipzyAuthException("Unauthorized");
        if (resp.statusCode() >= 400) throw new EpodClient.ShipzyException("HTTP " + resp.statusCode(), resp.statusCode());
        return gson.fromJson(resp.body(), t);
    }

    public Object listProviders(String capability) throws Exception {
        String q = capability != null ? "?capability=" + URLEncoder.encode(capability, StandardCharsets.UTF_8) : "";
        return request("/api/v1/marketplace/providers" + q, "GET", Object.class, null);
    }
    public Object getProvider(String slug) throws Exception { return request("/api/v1/marketplace/providers/" + slug, "GET", Object.class, null); }
    public Object listActivations(int page, int pageSize) throws Exception {
        return request("/api/v1/marketplace/activations?page=" + page + "&page_size=" + pageSize, "GET", Object.class, null);
    }
    public Object getActivation(String id) throws Exception { return request("/api/v1/marketplace/activations/" + id, "GET", Object.class, null); }
    public Object activate(Object data) throws Exception { return request("/api/v1/marketplace/activations", "POST", Object.class, data); }
    public Object pause(String id) throws Exception { return request("/api/v1/marketplace/activations/" + id + "/pause", "POST", Object.class, null); }
    public Object resume(String id) throws Exception { return request("/api/v1/marketplace/activations/" + id + "/resume", "POST", Object.class, null); }
    public Object revoke(String id, String reason) throws Exception { return request("/api/v1/marketplace/activations/" + id + "/revoke", "POST", Object.class, reason != null ? Map.of("reason", reason) : null); }
}

class TrackingClient {
    private final HttpClient httpClient;
    private final EpodClient.ShipzyConfig config;
    private static final Gson gson = new Gson();

    public TrackingClient(EpodClient.ShipzyConfig c) {
        this.config = c;
        this.httpClient = HttpClient.newBuilder().connectTimeout(Duration.ofSeconds(c.getTimeoutSeconds())).build();
    }
    public void setToken(String t) { config.setToken(t); }

    private <T> T request(String path, String method, Class<T> t, Object body) throws Exception {
        var builder = HttpRequest.newBuilder()
            .uri(URI.create(config.getBaseUrl().replaceAll("/$", "") + path))
            .header("Authorization", new EpodClient(config).getAuthHeader())
            .header("Content-Type", "application/json");
        if ("POST".equals(method) || "PUT".equals(method))
            builder.method(method, HttpRequest.BodyPublishers.ofString(body != null ? gson.toJson(body) : "{}"));
        else builder.GET();
        HttpResponse<String> resp = httpClient.send(builder.build(), HttpResponse.BodyHandlers.ofString());
        if (resp.statusCode() == 401) throw new EpodClient.ShipzyAuthException("Unauthorized");
        if (resp.statusCode() >= 400) throw new EpodClient.ShipzyException("HTTP " + resp.statusCode(), resp.statusCode());
        return gson.fromJson(resp.body(), t);
    }

    public Object detail(String trackingNo) throws Exception { return request("/api/v1/tracking/" + trackingNo, "GET", Object.class, null); }
    public Object list(int page, int pageSize, String status, String trackingNo) throws Exception {
        String basePath = EpodClient.UserRole.CARRIER.equals(config.getRole()) ? "/api/v1/carrier/tracking/list" : "/api/v1/merchant/tracking/list";
        StringBuilder q = new StringBuilder("?page=" + page + "&page_size=" + pageSize);
        if (status != null) q.append("&status=").append(URLEncoder.encode(status, StandardCharsets.UTF_8));
        if (trackingNo != null) q.append("&tracking_no=").append(URLEncoder.encode(trackingNo, StandardCharsets.UTF_8));
        return request(basePath + q, "GET", Object.class, null);
    }
}

class UploadClient {
    private final HttpClient httpClient;
    private final EpodClient.ShipzyConfig config;
    private static final Gson gson = new Gson();

    public UploadClient(EpodClient.ShipzyConfig c) {
        this.config = c;
        this.httpClient = HttpClient.newBuilder().connectTimeout(Duration.ofSeconds(c.getTimeoutSeconds())).build();
    }
    public void setToken(String t) { config.setToken(t); }

    private <T> T request(String path, String method, Class<T> t, Object body) throws Exception {
        var builder = HttpRequest.newBuilder()
            .uri(URI.create(config.getBaseUrl().replaceAll("/$", "") + path))
            .header("Authorization", new EpodClient(config).getAuthHeader())
            .header("Content-Type", "application/json");
        if ("POST".equals(method) || "PUT".equals(method))
            builder.method(method, HttpRequest.BodyPublishers.ofString(body != null ? gson.toJson(body) : "{}"));
        else builder.GET();
        HttpResponse<String> resp = httpClient.send(builder.build(), HttpResponse.BodyHandlers.ofString());
        if (resp.statusCode() == 401) throw new EpodClient.ShipzyAuthException("Unauthorized");
        if (resp.statusCode() >= 400) throw new EpodClient.ShipzyException("HTTP " + resp.statusCode(), resp.statusCode());
        return gson.fromJson(resp.body(), t);
    }

    public Object uploadFile(String endpoint, Object body) throws Exception { return request(endpoint, "POST", Object.class, body); }
    public Object brandingUploadLogo(Object body) throws Exception { return request("/api/v1/merchant/branding/logo", "POST", Object.class, body); }
}

class CarrierClient {
    private final HttpClient httpClient;
    private final EpodClient.ShipzyConfig config;
    private static final Gson gson = new Gson();

    public CarrierClient(EpodClient.ShipzyConfig c) {
        this.config = c;
        this.httpClient = HttpClient.newBuilder().connectTimeout(Duration.ofSeconds(c.getTimeoutSeconds())).build();
    }
    public void setToken(String t) { config.setToken(t); }

    private <T> T request(String path, String method, Class<T> t, Object body) throws Exception {
        var builder = HttpRequest.newBuilder()
            .uri(URI.create(config.getBaseUrl().replaceAll("/$", "") + path))
            .header("Authorization", new EpodClient(config).getAuthHeader())
            .header("Content-Type", "application/json");
        if ("POST".equals(method) || "PUT".equals(method) || "DELETE".equals(method))
            builder.method(method, HttpRequest.BodyPublishers.ofString(body != null ? gson.toJson(body) : "{}"));
        else builder.GET();
        HttpResponse<String> resp = httpClient.send(builder.build(), HttpResponse.BodyHandlers.ofString());
        if (resp.statusCode() == 401) throw new EpodClient.ShipzyAuthException("Unauthorized");
        if (resp.statusCode() >= 400) throw new EpodClient.ShipzyException("HTTP " + resp.statusCode(), resp.statusCode());
        return gson.fromJson(resp.body(), t);
    }

    public Object list(int page, int pageSize, String state) throws Exception {
        StringBuilder q = new StringBuilder("?page=" + page + "&page_size=" + pageSize);
        if (state != null) q.append("&state=").append(URLEncoder.encode(state, StandardCharsets.UTF_8));
        return request("/api/v1/carrier/list" + q, "GET", Object.class, null);
    }
    public Object get(String id) throws Exception { return request("/api/v1/carrier/" + id, "GET", Object.class, null); }
    public Object create(Object data) throws Exception { return request("/api/v1/carrier", "POST", Object.class, data); }
    public Object update(String id, Object data) throws Exception { return request("/api/v1/carrier/" + id, "PUT", Object.class, data); }
    public Object delete(String id) throws Exception { return request("/api/v1/carrier/" + id, "DELETE", Object.class, null); }
}

class PlatformConfigClient {
    private final HttpClient httpClient;
    private final EpodClient.ShipzyConfig config;
    private static final Gson gson = new Gson();

    public PlatformConfigClient(EpodClient.ShipzyConfig c) {
        this.config = c;
        this.httpClient = HttpClient.newBuilder().connectTimeout(Duration.ofSeconds(c.getTimeoutSeconds())).build();
    }
    public void setToken(String t) { config.setToken(t); }

    private <T> T request(String path, String method, Class<T> t, Object body) throws Exception {
        var builder = HttpRequest.newBuilder()
            .uri(URI.create(config.getBaseUrl().replaceAll("/$", "") + path))
            .header("Authorization", new EpodClient(config).getAuthHeader())
            .header("Content-Type", "application/json");
        if ("POST".equals(method) || "PUT".equals(method))
            builder.method(method, HttpRequest.BodyPublishers.ofString(body != null ? gson.toJson(body) : "{}"));
        else builder.GET();
        HttpResponse<String> resp = httpClient.send(builder.build(), HttpResponse.BodyHandlers.ofString());
        if (resp.statusCode() == 401) throw new EpodClient.ShipzyAuthException("Unauthorized");
        if (resp.statusCode() >= 400) throw new EpodClient.ShipzyException("HTTP " + resp.statusCode(), resp.statusCode());
        return gson.fromJson(resp.body(), t);
    }

    public Object list() throws Exception { return request("/api/v1/admin/platform-configs", "GET", Object.class, null); }
    public Object update(String id, Object data) throws Exception { return request("/api/v1/admin/platform-configs/" + id, "PUT", Object.class, data); }
}

class ComplianceClient {
    private final HttpClient httpClient;
    private final EpodClient.ShipzyConfig config;
    private static final Gson gson = new Gson();

    public ComplianceClient(EpodClient.ShipzyConfig c) {
        this.config = c;
        this.httpClient = HttpClient.newBuilder().connectTimeout(Duration.ofSeconds(c.getTimeoutSeconds())).build();
    }
    public void setToken(String t) { config.setToken(t); }

    private <T> T request(String path, String method, Class<T> t, Object body) throws Exception {
        var builder = HttpRequest.newBuilder()
            .uri(URI.create(config.getBaseUrl().replaceAll("/$", "") + path))
            .header("Authorization", new EpodClient(config).getAuthHeader())
            .header("Content-Type", "application/json");
        if ("POST".equals(method) || "PUT".equals(method))
            builder.method(method, HttpRequest.BodyPublishers.ofString(body != null ? gson.toJson(body) : "{}"));
        else builder.GET();
        HttpResponse<String> resp = httpClient.send(builder.build(), HttpResponse.BodyHandlers.ofString());
        if (resp.statusCode() == 401) throw new EpodClient.ShipzyAuthException("Unauthorized");
        if (resp.statusCode() >= 400) throw new EpodClient.ShipzyException("HTTP " + resp.statusCode(), resp.statusCode());
        return gson.fromJson(resp.body(), t);
    }

    public Object check(Object data) throws Exception { return request("/api/v1/compliance/check", "POST", Object.class, data); }
    public Object countryRequirements(String countryCode) throws Exception { return request("/api/v1/compliance/requirements/" + countryCode, "GET", Object.class, null); }
    public Object createCustoms(Object data) throws Exception { return request("/api/v1/compliance/customs", "POST", Object.class, data); }
    public Object getCustoms(String id) throws Exception { return request("/api/v1/compliance/customs/" + id, "GET", Object.class, null); }
    public Object validateHsCode(String hsCode) throws Exception { return request("/api/v1/compliance/hscode/" + hsCode + "/validate", "GET", Object.class, null); }
    public Object prohibitedItems() throws Exception { return request("/api/v1/compliance/prohibited", "GET", Object.class, null); }
}

class CpscClient {
    private final HttpClient httpClient;
    private final EpodClient.ShipzyConfig config;
    private static final Gson gson = new Gson();

    public CpscClient(EpodClient.ShipzyConfig c) {
        this.config = c;
        this.httpClient = HttpClient.newBuilder().connectTimeout(Duration.ofSeconds(c.getTimeoutSeconds())).build();
    }
    public void setToken(String t) { config.setToken(t); }

    private <T> T request(String path, String method, Class<T> t, Object body) throws Exception {
        var builder = HttpRequest.newBuilder()
            .uri(URI.create(config.getBaseUrl().replaceAll("/$", "") + path))
            .header("Authorization", new EpodClient(config).getAuthHeader())
            .header("Content-Type", "application/json");
        if ("POST".equals(method) || "PUT".equals(method))
            builder.method(method, HttpRequest.BodyPublishers.ofString(body != null ? gson.toJson(body) : "{}"));
        else builder.GET();
        HttpResponse<String> resp = httpClient.send(builder.build(), HttpResponse.BodyHandlers.ofString());
        if (resp.statusCode() == 401) throw new EpodClient.ShipzyAuthException("Unauthorized");
        if (resp.statusCode() >= 400) throw new EpodClient.ShipzyException("HTTP " + resp.statusCode(), resp.statusCode());
        return gson.fromJson(resp.body(), t);
    }

    public Object getCollections() throws Exception { return request("/api/v1/cpsc/collections", "GET", Object.class, null); }
    public Object getCredential() throws Exception { return request("/api/v1/cpsc/credential", "GET", Object.class, null); }
    public Object saveCredential(Object data) throws Exception { return request("/api/v1/cpsc/credential", "POST", Object.class, data); }
    public Object importData(Object data) throws Exception { return request("/api/v1/cpsc/import", "POST", Object.class, data); }
    public Object getImportStatus(String importId) throws Exception { return request("/api/v1/cpsc/import/" + importId + "/status", "GET", Object.class, null); }
    public Object getImportLog(String importId, boolean errorsOnly) throws Exception { return request("/api/v1/cpsc/import/" + importId + "/log?errorsOnly=" + errorsOnly, "GET", Object.class, null); }
    public Object exportData(Map<String, Object> filter) throws Exception {
        StringBuilder q = new StringBuilder();
        if (filter != null) filter.forEach((k, v) -> q.append("&").append(k).append("=").append(v));
        return request("/api/v1/cpsc/export" + q, "GET", Object.class, null);
    }
    public Object exportAsync(Map<String, Object> filter) throws Exception {
        StringBuilder q = new StringBuilder();
        if (filter != null) filter.forEach((k, v) -> q.append("&").append(k).append("=").append(v));
        return request("/api/v1/cpsc/export-async" + q, "GET", Object.class, null);
    }
    public Object getExportAsyncStatus(String exportId) throws Exception { return request("/api/v1/cpsc/export-async/" + exportId + "/status", "GET", Object.class, null); }
    public Object getExportAsyncData(String exportId) throws Exception { return request("/api/v1/cpsc/export-async/" + exportId + "/data", "GET", Object.class, null); }
    public Object getCertificates(Object data) throws Exception { return request("/api/v1/cpsc/certificates", "POST", Object.class, data); }
    public Object getTradeParties(String partyType) throws Exception {
        String q = partyType != null ? "?tradePartyType=" + URLEncoder.encode(partyType, StandardCharsets.UTF_8) : "";
        return request("/api/v1/cpsc/trade-parties" + q, "GET", Object.class, null);
    }
    public Object getTokenExpiration() throws Exception { return request("/api/v1/cpsc/token-expiration", "GET", Object.class, null); }
}

class ShipzyClient {
    public final EpodClient epod;
    public final OrderClient order;
    public final AddressClient address;
    public final CarrierEpodClient carrierEpod;
    public final EcmrClient ecmr;
    public final ProductClient product;
    public final ActivationClient activation;
    public final TrackingClient tracking;
    public final UploadClient upload;
    public final CarrierClient carrier;
    public final PlatformConfigClient platformConfig;
    public final ComplianceClient compliance;
    public final CpscClient cpsc;
    public final EpodClient.UserRole role;

    public ShipzyClient(EpodClient.ShipzyConfig c) {
        epod = new EpodClient(c);
        order = new OrderClient(c);
        address = new AddressClient(c);
        carrierEpod = new CarrierEpodClient(c);
        ecmr = new EcmrClient(c);
        product = new ProductClient(c);
        activation = new ActivationClient(c);
        tracking = new TrackingClient(c);
        upload = new UploadClient(c);
        carrier = new CarrierClient(c);
        platformConfig = new PlatformConfigClient(c);
        compliance = new ComplianceClient(c);
        cpsc = new CpscClient(c);
        role = c.getRole();
    }

    public void updateToken(String t) {
        epod.setToken(t); order.setToken(t); address.setToken(t);
        carrierEpod.setToken(t); ecmr.setToken(t); product.setToken(t);
        activation.setToken(t); tracking.setToken(t); upload.setToken(t);
        carrier.setToken(t); platformConfig.setToken(t); compliance.setToken(t); cpsc.setToken(t);
    }

    public boolean isMerchant() { return role == EpodClient.UserRole.MERCHANT; }
    public boolean isCarrier() { return role == EpodClient.UserRole.CARRIER; }
}

File: kotlin/src/main/kotlin/com/shipzy/sdk/EpodClient.kt
package com.shipzy.sdk

import io.ktor.client.*
import io.ktor.client.call.*
import io.ktor.client.engine.cio.*
import io.ktor.client.plugins.*
import io.ktor.client.request.*
import io.ktor.http.*
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

enum class UserRole { MERCHANT, CARRIER }

data class ShipzyConfig(
    val baseUrl: String = "https://api.shipzy.me",
    var token: String = "",
    val timeoutSeconds: Int = 30,
    var role: UserRole = UserRole.MERCHANT,
    var carrierCode: String = ""
)

@Serializable
data class EpodListResponse(val data: List<EpodListItem> = emptyList(), val total: Int = 0, val page: Int = 0, @SerialName("page_size") val pageSize: Int = 0)
@Serializable
data class EpodListItem(val id: String = "", @SerialName("tracking_no") val trackingNo: String = "", val status: String = "", @SerialName("recipient_name") val recipientName: String? = null, @SerialName("created_at") val createdAt: String = "")
@Serializable
data class EpodDetail(val id: String = "", @SerialName("tracking_no") val trackingNo: String = "", val status: String = "", @SerialName("recipient_name") val recipientName: String? = null, @SerialName("recipient_phone") val recipientPhone: String? = null, @SerialName("created_at") val createdAt: String = "", @SerialName("updated_at") val updatedAt: String = "", @SerialName("sign_url") val signUrl: String? = null, @SerialName("evidence_hash") val evidenceHash: String? = null)
@Serializable
data class SignUrlResponse(@SerialName("sign_url") val signUrl: String = "")

open class ShipzyException(message: String, val statusCode: Int) : Exception(message)
class ShipzyAuthException(message: String) : ShipzyException(message, 401)

class EpodClient(private val config: ShipzyConfig) {
    private val json = Json { ignoreUnknownKeys = true; isLenient = true }
    private val client = HttpClient(CIO) { install(HttpTimeout) { requestTimeoutMillis = config.timeoutSeconds * 1000L } }
    fun setToken(token: String) { config.token = token }
    private fun getAuthHeader(): String = if (config.role == UserRole.CARRIER && config.carrierCode.isNotEmpty()) "Bearer ${config.carrierCode}:${config.token}" else "Bearer ${config.token}"
    private suspend inline fun <reified T> request(path: String, method: HttpMethod = HttpMethod.Get, body: Any? = null): T {
        val response = client.request(config.baseUrl.trimEnd('/') + path) {
            this.method = method; header(HttpHeaders.Authorization, getAuthHeader()); header(HttpHeaders.ContentType, ContentType.Application.Json)
            if (body != null) setBody(body)
        }
        if (response.status == HttpStatusCode.Unauthorized) throw ShipzyAuthException("Unauthorized")
        if (!response.status.isSuccess()) throw ShipzyException("HTTP ${response.status.value}", response.status.value)
        return response.body()
    }
    suspend fun list(page: Int = 1, pageSize: Int = 25, status: String? = null, trackingNo: String? = null): EpodListResponse {
        val q = listOfNotNull("page=$page", "page_size=$pageSize", status?.let { "status=${java.net.URLEncoder.encode(it, "UTF-8")}" }, trackingNo?.let { "tracking_no=${java.net.URLEncoder.encode(it, "UTF-8")}" }).joinToString("&")
        return request("/api/v1/shipment/epod/list?$q")
    }
    suspend fun get(id: String): EpodDetail = request("/api/v1/shipment/epod/$id")
    suspend fun create(data: Any): EpodDetail = request("/api/v1/shipment/epod/create", HttpMethod.Post, data)
    suspend fun generateFromOrder(orderId: String): EpodDetail = request("/api/v1/shipment/epod/generate-from-order", HttpMethod.Post, mapOf("order_id" to orderId))
    suspend fun update(id: String, data: Any): EpodDetail = request("/api/v1/shipment/epod/$id/update", HttpMethod.Put, data)
    suspend fun deliver(id: String): EpodDetail = request("/api/v1/shipment/epod/$id/delivery", HttpMethod.Post)
    suspend fun fail(id: String, remark: String): EpodDetail = request("/api/v1/shipment/epod/$id/fail", HttpMethod.Post, mapOf("remark" to remark))
    suspend fun generateSignUrl(id: String): SignUrlResponse = request("/api/v1/shipment/epod/$id/sign", HttpMethod.Post)
}

class OrderClient(private val config: ShipzyConfig) {
    private val client = HttpClient(CIO) { install(HttpTimeout) { requestTimeoutMillis = config.timeoutSeconds * 1000L } }
    fun setToken(token: String) { config.token = token }
    private suspend inline fun <reified T> request(path: String, method: HttpMethod = HttpMethod.Get, body: Any? = null): T {
        val response = client.request(config.baseUrl.trimEnd('/') + path) { this.method = method; header(HttpHeaders.Authorization, "Bearer ${config.token}"); header(HttpHeaders.ContentType, ContentType.Application.Json); if (body != null) setBody(body) }
        if (response.status == HttpStatusCode.Unauthorized) throw ShipzyAuthException("Unauthorized")
        if (!response.status.isSuccess()) throw ShipzyException("HTTP ${response.status.value}", response.status.value)
        return response.body()
    }
    suspend fun list(page: Int = 1, pageSize: Int = 25, status: String? = null): EpodListResponse {
        val q = listOfNotNull("page=$page", "page_size=$pageSize", status?.let { "status=${java.net.URLEncoder.encode(it, "UTF-8")}" }).joinToString("&")
        return request("/api/v1/order/list?$q")
    }
    suspend fun get(id: String): Any = request("/api/v1/order/$id")
    suspend fun create(data: Any): Any = request("/api/v1/order/create", HttpMethod.Post, data)
    suspend fun update(id: String, data: Any): Any = request("/api/v1/order/$id/update", HttpMethod.Post, data)
    suspend fun cancel(id: String): Any = request("/api/v1/order/$id/cancel", HttpMethod.Post)
}

class AddressClient(private val config: ShipzyConfig) {
    private val client = HttpClient(CIO) { install(HttpTimeout) { requestTimeoutMillis = config.timeoutSeconds * 1000L } }
    fun setToken(token: String) { config.token = token }
    private suspend inline fun <reified T> request(path: String, method: HttpMethod = HttpMethod.Post, body: Any? = null): T {
        val response = client.request(config.baseUrl.trimEnd('/') + path) { this.method = method; header(HttpHeaders.Authorization, "Bearer ${config.token}"); header(HttpHeaders.ContentType, ContentType.Application.Json); if (body != null) setBody(body) }
        if (response.status == HttpStatusCode.Unauthorized) throw ShipzyAuthException("Unauthorized")
        if (!response.status.isSuccess()) throw ShipzyException("HTTP ${response.status.value}", response.status.value)
        return response.body()
    }
    suspend fun list(params: Any? = null): Any = request("/api/v1/merchant/addresses/list", body = params)
    suspend fun create(data: Any): Any = request("/api/v1/merchant/addresses/create", body = data)
    suspend fun update(id: String, data: Any): Any = request("/api/v1/merchant/addresses/$id/update", body = data)
    suspend fun delete(id: String): Any = request("/api/v1/merchant/addresses/$id/delete")
}

class CarrierEpodClient(private val config: ShipzyConfig) {
    private val client = HttpClient(CIO) { install(HttpTimeout) { requestTimeoutMillis = config.timeoutSeconds * 1000L } }
    fun setToken(token: String) { config.token = token }
    private suspend inline fun <reified T> request(path: String, method: HttpMethod = HttpMethod.Get, body: Any? = null): T {
        val response = client.request(config.baseUrl.trimEnd('/') + path) { this.method = method; header(HttpHeaders.Authorization, "Bearer ${config.token}"); header(HttpHeaders.ContentType, ContentType.Application.Json); if (body != null) setBody(body) }
        if (response.status == HttpStatusCode.Unauthorized) throw ShipzyAuthException("Unauthorized")
        if (!response.status.isSuccess()) throw ShipzyException("HTTP ${response.status.value}", response.status.value)
        return response.body()
    }
    suspend fun list(page: Int = 1, pageSize: Int = 25, status: String? = null): EpodListResponse {
        val q = listOfNotNull("page=$page", "page_size=$pageSize", status?.let { "status=${java.net.URLEncoder.encode(it, "UTF-8")}" }).joinToString("&")
        return request("/api/v1/carrier/epod/list?$q")
    }
    suspend fun get(id: String): EpodDetail = request("/api/v1/carrier/epod/$id")
    suspend fun deliver(id: String): EpodDetail = request("/api/v1/carrier/epod/$id/delivery", HttpMethod.Post)
    suspend fun fail(id: String, remark: String): EpodDetail = request("/api/v1/carrier/epod/$id/fail", HttpMethod.Post, mapOf("remark" to remark))
}

class ShipzyClient(config: ShipzyConfig) {
    val epod = EpodClient(config)
    val order = OrderClient(config)
    val address = AddressClient(config)
    val carrierEpod = CarrierEpodClient(config)
    val validation = ValidationClient(config)
    val role = config.role
    fun updateToken(token: String) {
        epod.setToken(token); order.setToken(token); address.setToken(token)
        carrierEpod.setToken(token); validation.setToken(token)
    }
    fun isMerchant() = role == UserRole.MERCHANT
    fun isCarrier() = role == UserRole.CARRIER
}

class ValidationClient(private val config: ShipzyConfig) {
    private val client = HttpClient(CIO) { install(HttpTimeout) { requestTimeoutMillis = config.timeoutSeconds * 1000L } }
    fun setToken(token: String) { config.token = token }
    private suspend inline fun <reified T> request(path: String, method: HttpMethod = HttpMethod.Post, body: Any? = null): T {
        val response = client.request(config.baseUrl.trimEnd('/') + path) {
            this.method = method; header(HttpHeaders.Authorization, "Bearer ${config.token}"); header(HttpHeaders.ContentType, ContentType.Application.Json)
            if (body != null) setBody(body)
        }
        if (response.status == HttpStatusCode.Unauthorized) throw ShipzyAuthException("Unauthorized")
        if (!response.status.isSuccess()) throw ShipzyException("HTTP ${response.status.value}", response.status.value)
        return response.body()
    }
    suspend fun verifyPhone(countryCode: String, phone: String): Map<String, Any> =
        request("/api/v1/validation/phone", HttpMethod.Post, mapOf("country_code" to countryCode, "phone" to phone))
    suspend fun formatPhone(countryCode: String, phone: String): Map<String, Any> =
        request("/api/v1/validation/phone/format", HttpMethod.Post, mapOf("country_code" to countryCode, "phone" to phone))
    suspend fun validatePostalCode(countryCode: String, code: String): Map<String, Any> =
        request("/api/v1/validation/postal-code", HttpMethod.Post, mapOf("country_code" to countryCode, "code" to code))
    suspend fun validateEmail(email: String): Map<String, Any> =
        request("/api/v1/validation/email", HttpMethod.Post, mapOf("email" to email))
    suspend fun validateTaxId(countryCode: String, taxId: String): Map<String, Any> =
        request("/api/v1/validation/tax-id", HttpMethod.Post, mapOf("country_code" to countryCode, "tax_id" to taxId))
    suspend fun health(): Map<String, Any> =
        request("/api/v1/validation/health", HttpMethod.Get)
}
